### PR TITLE
Ability to provide custom configuration with --config flag

### PR DESCRIPTION
**Pull request by samirsilwal**
## Changes
- Added `--config` flag in each command to provide custom config source for each command run
Screenshot from 2021-03-26 14-21-27
Screenshot from 2021-03-26 14-21-50
Screenshot from 2021-03-26 14-22-32
Screenshot from 2021-03-26 14-22-52

### CODE REVIEW

File: test/cli/commands/make-publish.test.ts [hunk 1]
```diff
@@ -143,4 +143,63 @@ describe('CLI: make-publish', () => {
     expect(createUp).to.equal('CREATE TABLE {{table}} (id INT PRIMARY KEY)');
     expect(createDown).to.equal('DROP TABLE {{table}}');
   });
+
+  it('should create respective stub files based on configuration from --config flag.', async () => {
```
Comment (samirsilwal) [inline test/cli/commands/make-publish.test.ts:147]:
```suggestion
  it('Should create respective stub files based on configuration from --config flag.', async () => {
```

File: test/cli/commands/make.test.ts [hunk 2]
```diff
@@ -334,4 +301,55 @@ describe('CLI: make', () => {
     expect(upFile).to.equal(interpolate(upSQL, { table: 'settings' }));
     expect(downFile).to.equal(interpolate(downSQL, { table: 'settings' }));
   });
+
+  it('should make migration based on custom configurations with --config flag.', async () => {
```
Comment (samirsilwal) [inline test/cli/commands/make.test.ts:305]:
```suggestion
  it('Should make migration based on custom configurations with --config flag.', async () => {
```

**Review by aanchal-adk — APPROVED** [review]

File: src/config.ts [hunk 3]
```diff
@@ -41,10 +41,10 @@ export function getSqlBasePath(config: Configuration): string {
  *
  * @returns {Promise<Configuration>}
  */
-export async function loadConfig(): Promise<Configuration> {
+export async function loadConfig(configFileName: string = CONFIG_FILENAME): Promise<Configuration> {
```
Comment (cham11ng) [inline src/config.ts:44]:
```suggestion
export async function loadConfig(configFilename: string = CONFIG_FILENAME): Promise<Configuration> {
```

File: src/commands/synchronize.ts [hunk 4]
```diff
@@ -25,6 +25,11 @@ class Synchronize extends Command {
     'connection-resolver': flags.string({
       helpValue: 'PATH',
       description: 'Path to the connection resolver.'
+    }),
+    config: flags.string({
+      char: 'c',
+      default: 'sync-db.yml',
```
Comment (cham11ng) [inline src/commands/synchronize.ts:31]:
```suggestion
      default: CONFIG_FILENAME,
```

File: src/config.ts [hunk 5]
```diff
@@ -41,10 +41,10 @@ export function getSqlBasePath(config: Configuration): string {
  *
  * @returns {Promise<Configuration>}
  */
-export async function loadConfig(): Promise<Configuration> {
+export async function loadConfig(configFilename: string = CONFIG_FILENAME): Promise<Configuration> {
   log('Resolving config file.');
 
-  const filename = path.resolve(process.cwd(), CONFIG_FILENAME);
+  const filename = path.resolve(process.cwd(), configFilename);
```
Comment (cham11ng) [inline src/config.ts:47]: | Let's make sure all sync-db configurations are prefixed with `sync-db-*`.

https://github.com/leapfrogtechnology/sync-db/pull/119#issuecomment-807996558

Should we throw an exception if this configuration filename pattern doesn't match?

cc @mesaugat

File: package.json [hunk 6]
```diff
@@ -1,7 +1,7 @@
 {
   "name": "@leapfrogtechnology/sync-db",
   "description": "Command line utility to synchronize and version control relational database objects across databases",
-  "version": "1.0.0-beta.11",
+  "version": "1.0.0-beta.12",
```
Comment (mesaugat) [inline package.json:4]: Let's release separately. Not in the same PR. @cham11ng @aanchal-adk @samirsilwal

Comment (mesaugat) [reply package.json:4]: We are abusing this ability.

File: src/commands/make-publish.ts [hunk 7]
```diff
@@ -1,20 +1,28 @@
...
 import { printInfo, printLine } from '../util/io';
 import * as fileMakerService from '../service/fileMaker';
 
 class MakePublish extends Command {
   static description = 'Publish migration templates files.';
-
+  static flags = {
+    config: flags.string({
+      char: 'c',
+      default: CONFIG_FILENAME,
```
Comment (mesaugat) [inline src/commands/make-publish.ts:14]: Shouldn't have any defaults. We should not run the command if `-c` is run without providing the configuration file name.

File: src/commands/make-publish.ts [hunk 8]
```diff
@@ -1,20 +1,28 @@
...
+      description: 'Custom configuration file.'
+    })
+  };
   /**
```
Comment (mesaugat) [inline src/commands/make-publish.ts:18]: Add a new line above this.

File: src/config.ts (hunk 5 above)
Comment (mesaugat) [reply src/config.ts:47]: Yes we should. Please limit it to `sync-db.yml` and `sync-db-*.yml` file names.

File: test/cli/commands/make.test.ts (hunk 2 above)
Comment (mesaugat) [inline test/cli/commands/make.test.ts:305]: Can we add a few more tests in `config.test.ts` to check the contents of the config file only? 

`should load the config file that is provided`
`should load the config file only if it matches the naming convention`
`should throw an error if the config file doesn't match the naming convention`

Comment (cham11ng) [reply test/cli/commands/make.test.ts:305]: should load the config file (absolute path)

File: src/config.ts (hunk 5 above)
Comment (cham11ng) [reply src/config.ts:47]: const filename = path.isAbsolute(configFilename)
    ? configFilename
    : path.join(process.cwd(), configFilename);

File: src/config.ts [hunk 9]
```diff
@@ -11,6 +11,8 @@ import ConnectionsFileSchema from './domain/ConnectionsFileSchema';
 import { prepareInjectionConfigVars } from './service/configInjection';
 import { DEFAULT_CONFIG, CONFIG_FILENAME, CONNECTIONS_FILENAME, REQUIRED_ENV_KEYS } from './constants';
 
+const CONFIG_FILE_CONVENTION = /^sync-db(-?)[\w]*\.yml$/;
```
Comment (mesaugat) [inline src/config.ts:14]: This is probably going to match these as well. We can so a simple check 

Check `^sync-db-[\w]+\.yml$` and `sync-db.yml` separately.

```
sync-db-.yml
sync-dbasdfafd.yml
```

File: test/unit/util/config.test.ts [hunk 10]
```diff
@@ -109,4 +113,114 @@ describe('CONFIG:', () => {
...
+        Error,
+        `The config filename doesn't meet the pattern (sync-db.yml or sync-db-*.yml)`
+      );
+      await expect(loadConfig('sync-db-test.yml')).not.to.be.rejectedWith(
+        Error,
+        `The config filename doesn't meet the pattern (sync-db.yml or sync-db-*.yml)`
+      );
+
+      const config = await loadConfig();
+      expect(config).to.have.property('migration');
```
Comment (mesaugat) [inline test/unit/util/config.test.ts:183]: No need to expect this here? Let's try to expect only one thing from one test.

File: test/unit/util/config.test.ts [hunk 11]
```diff
@@ -109,4 +113,114 @@ describe('CONFIG:', () => {
...
+
+      const config1 = await loadConfig('sync-db-test.yml');
+      expect(config1).to.have.property('migration');
```
Comment (mesaugat) [inline test/unit/util/config.test.ts:186]: .

File: test/unit/util/config.test.ts [hunk 12]
```diff
@@ -109,4 +113,114 @@ describe('CONFIG:', () => {
...
+      );
+      await expect(loadConfig('sync-db-test.js')).to.be.rejectedWith(
+        Error,
+        `The config filename doesn't meet the pattern (sync-db.yml or sync-db-*.yml)`
+      );
+      await expect(loadConfig('sync.yml')).to.be.rejectedWith(
+        Error,
+        `The config filename doesn't meet the pattern (sync-db.yml or sync-db-*.yml)`
+      );
+      await expect(loadConfig('sync-db.yml.txt')).to.be.rejectedWith(
```
Comment (mesaugat) [inline test/unit/util/config.test.ts:202]: Add tests for `sync-db-.yml` and `sync-dbasdfasdf.yml`.

**Review by mesaugat — APPROVED** [review]
Looks good to me.

**Pull request merged**